[Frontend] Add experimental flag to skip non-inlinable function bodies

harlanhaskins opened the pull request:
This patch adds an experimental flag to skip non-inlinable function
bodies and turns it on for the standard library and overlay
`.swiftmodules`.

This flag is disabled if the requested action generates IR.

---

This option causes some small improvements in the time it takes to type check and generate an optimized and unoptimized standard library `.swiftmodule`

| Mode                                 | Optimization | Time     | Time (Skipping) | Ratio |
|:-------------------------------------|:-------------|:---------|:----------------|:------|
| `-emit-parseable-module-interface`   | N/A          | 16.254s  | 14.947s         | 1.09x |
| `-emit-module`                       | `-Onone`     | 21.685s  | 19.462s         | 1.11x |
| `-emit-module`                       | `-O`         | 50.526s  | 35.234s         | 1.43x |

However, since the standard library is full of inlinable code, it doesn't benefit much from skipping non-inlinable code.

To find a more representative Swift project with little inlinable code, I ran these benchmarks against the current master of SwiftLint, which shows much stronger wins:

| Mode                                 | Optimization | Time     | Time (Skipping) | Ratio     |
|:-------------------------------------|:-------------|:---------|:----------------|:----------|
| `-emit-parseable-module-interface`   | N/A          | 13.179s  | 1.732s          | **7.70x** |
| `-emit-module`                       | `-Onone`     | 16.21s   | 2.719s          | **5.88x** |
| `-emit-module`                       | `-O`         | 63.408s  | 12.186s         | **5.26x** |

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
Okay, looks like I’m missing some specializations or some inlinable code.

harlanhaskins commented:
The Linux failures might shed some light on what I'm missing:

```
Test Result (4 failures / +3)
  - Swift(linux-x86_64).SILOptimizer.static_arrays.swift
  - Swift(linux-x86_64).SILOptimizer.optionset.swift
  - Swift(linux-x86_64).SILOptimizer.pointer_conversion.swift
  - Swift(linux-x86_64).SILOptimizer.array_contentof_opt.swift
```

harlanhaskins commented:
Oops! Was skipping inlinable deinitializers.

@swift-ci please benchmark

harlanhaskins commented:
Still missing _something_...

harlanhaskins commented:
@swift-ci please smoke test linux platform

harlanhaskins commented:
Running these again given the @inline(__always) changes.

@swift-ci please benchmark

harlanhaskins commented:
Just gonna kick off another benchmarking run with @jrose-apple's recent pointer-related changes in #21126.

@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
Seems that wasn't it.

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
Gotta remember to leave SwiftOnoneSupport alone...

@swift-ci please benchmark

harlanhaskins commented:
Getting much closer!

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -425,7 +425,10 @@ class alignas(1 << DeclAlignInBits) Decl {
+    HasSingleExpressionBody : 1,
+
+    /// Whether the function body was skipped during typechecking.
+    WasBodyTypecheckingSkipped : 1
```
The existing BodyKind should capture this already

slavapestov commented on lib/AST/DeclContext.cpp:
```diff
@@ -340,10 +340,16 @@ swift::ResilienceExpansionRequest::evaluate(Evaluator &evaluator,
-    if (isa<PatternBindingInitializer>(dc)) {
-      if (auto *NTD = dyn_cast<NominalTypeDecl>(dc->getParent())) {
+    // Non-static stored property initializer contexts use minimal resilience
+    // expansion if the type is formally fixed layout.
```
This looks like an unrelated change, it warrants its own commit and tests

slavapestov commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2139,13 +2139,25 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
+      !AFD->getDeclContext()->isLocalContext() &&
+      AFD->getResilienceExpansion() == ResilienceExpansion::Maximal) {
+    AFD->setBodyTypecheckingSkipped(true);
+    return false;
```
It would be better if we just didn't add these functions to TypeChecker::definedFunctions. We shouldn't get to running this request on them at all

slavapestov commented on lib/Sema/TypeCheckError.cpp:
```diff
@@ -1627,6 +1627,10 @@ void TypeChecker::checkFunctionErrorHandling(AbstractFunctionDecl *fn) {
 
+  // If we skipped typechecking the body, don't perform this check; it will
+  // always warn.
+  if (fn->wasBodyTypecheckingSkipped()) return;
```
Would be better if checkFunctionErrorHandling() didn't get called on these functions at all

slavapestov commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2139,13 +2139,25 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
+  // still do typecheck functions that appear inside a local context, because
+  // their parent context will have been typechecked.
+  if (tc.SkipNonInlinableFunctionBodies &&
+      !AFD->getDeclContext()->isLocalContext() &&
```
Why is there an exception for local functions?

slavapestov commented on lib/Serialization/Serialization.cpp:
```diff
@@ -4909,6 +4909,12 @@ void Serializer::writeAST(ModuleOrSourceFile DC,
       Mangle::ASTMangler Mangler;
+
+      // If we skipped the parent of this type, don't bother serializing it.
+      if (auto afd = dyn_cast<AbstractFunctionDecl>(TD->getDeclContext()))
```
A local type can appear inside other local contexts, such as a property initializer or a closure

slavapestov commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -400,6 +400,13 @@ void swift::performTypeChecking(SourceFile &SF, TopLevelContext &TLC,
       TC.setInImmediateMode(true);
 
+    if (Options.contains(TypeCheckingFlags::SkipNonInlinableFunctionBodies))
+      // Disable this optimization if we're compiling SwiftOnoneSupport, because
```
We shouldn't be generating a swiftinterface for SwiftOnoneSupport, though?

harlanhaskins replied:
This also applies to -emit-module

slavapestov commented on lib/SILGen/SILGen.cpp:
```diff
@@ -803,10 +803,13 @@ emitMarkFunctionEscapeForTopLevelCodeGlobals(SILLocation loc,
   // Emit any default argument generators.
   emitDefaultArgGenerators(AFD, AFD->getParameters());
 
+  if (AFD->wasBodyTypecheckingSkipped())
```
Instead, I would prefer it if the skipping was done further up the callstack. Returning a boolean is subtle

slavapestov replied:
... or further down. See how emitOrDelayFunction takes a SILDeclRef and a callback? If the SILDeclRef's isSerialized() is not set, you can just not call the callback and return. This will handle default args, accessors, functions, and most other things, except for vtable and witness thunks.

slavapestov reviewed (commented):
Cool! You should investigate enabling delayed parsing for primary files in this mode as well, so that we don't even have to build AST for their bodies.

slavapestov commented:
Even more things to investigate:

- There are parts of SILGenType you can look into skipping as well, for example emitting serialized vtables and witness tables.
- Skipping global variables that do not have inlinable accessors.
- To check your work: add a SIL pass that asserts that when the flag is set, there aren't any non-serializable SILFunctions in the final SILModule.

slavapestov commented on lib/SILGen/SILGen.cpp:
```diff
@@ -803,10 +803,13 @@ emitMarkFunctionEscapeForTopLevelCodeGlobals(SILLocation loc,
-void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
+bool SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
   // Emit any default argument generators.
   emitDefaultArgGenerators(AFD, AFD->getParameters());
```
Will this still emit default arg initializers for private/internal functions?

harlanhaskins commented on include/swift/AST/Decl.h:
```diff
@@ -5714,14 +5714,17 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   /// cannot be attached after this call.
   void setBodySkipped(SourceRange bodyRange) {
-    assert(getBodyKind() == BodyKind::None);
+//    assert(getBodyKind() == BodyKind::None ||
```
This needs a FIXME, I'd like to change the assert to the commented version once we enable delayed parsing everywhere.

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please benchmark

harlanhaskins commented:
This benchmark isn't actually testing the feature, I just want to make sure I didn't regress anything when the flag isn't set.

@swift-ci please benchmark

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
Python lint failure

@swift-ci please smoke test macOS

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test macOS

harlanhaskins commented:
@swift-ci please smoke test

CodaFi commented on lib/SILGen/SILGen.cpp:
```diff
@@ -1644,6 +1644,22 @@ class SourceFileScope {
+// Determines if the provided type decl is embedded at some point in an
+// AbstractFunctionDecl whose body has been skipped.
+// FIXME: Duplicated from lib/Serialization/Serialization.cpp
+static bool isInSkippedFunctionContext(const TypeDecl *td) {
```
This might be good to lift into libAST

CodaFi commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1835,82 +1835,6 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
 }
 
-static Optional<unsigned>
-getParamIndex(const ParameterList *paramList, const ParamDecl *decl) {
```
Can this change be made in isolation?

CodaFi commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2145,7 +2069,6 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
 
   // FIXME(InterfaceTypeRequest): Remove this.
   (void)AFD->getInterfaceType();
```
You should be able to get rid of this too.

harlanhaskins replied:
Tried it, there are a couple of code completion issues, likely a missing forced interface type. Leaving for a future PR.

harlanhaskins committed:
Convert TypeChecker::checkDefaultArguments to static function

This doesn’t need to be a method, since it’s now only going to be used in the Decl checker

harlanhaskins committed:
[Modules] Add flag to skip non-inlinable function bodies

This flag, currently staged in as `-experimental-skip-non-inlinable-function-bodies`, will cause the typechecker to skip typechecking bodies of functions that will not be serialized in the resulting `.swiftmodule`. This patch also includes a SIL verifier that ensures that we don’t accidentally include a body that we should have skipped.

There is still some work left to make sure the emitted .swiftmodule is exactly the same as what’s emitted without the flag, which is what’s causing the benchmark noise above. I’ll be committing follow-up patches to address those, but for now I’m going to land the implementation behind a flag.

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -254,6 +254,10 @@ def stats_output_dir: Separate<["-"], "stats-output-dir">,
+def experimental_skip_non_inlinable_function_bodies:
+  Flag<["-"], "experimental-skip-non-inlinable-function-bodies">,
+  Flags<[FrontendOption, HelpHidden]>,
+  HelpText<"Skip type-checking and SIL generation for non-inlinable function bodies">;
```
Nitpick: why is this in the middle of all the stats options?

(Also, why is it a Driver option again?)

jrose-apple commented on lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp:
```diff
@@ -0,0 +1,107 @@
+  // FIXME: We can probably skip property initializers, too.
+  auto func = F.getLocation().getAsASTNode<AbstractFunctionDecl>();
+  if (!func)
+    return false;
```
Yeah, this scares me. All sorts of synthesized things will slip through here too.

jrose-apple commented on lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp:
```diff
@@ -0,0 +1,107 @@
+
+  // Local function bodies in inlinable code are okay to show up in the module.
+  if (func->getDeclContext()->isLocalContext())
+    return false;
```
Worth checking that the enclosing context is in fact inlinable.

jrose-apple commented on lib/Sema/TypeChecker.h:
```diff
@@ -707,6 +711,14 @@ class TypeChecker final : public LazyResolver {
+
+  bool canSkipNonInlinableBodies() const {
+    return SkipNonInlinableFunctionBodies;
+  }
```
Nitpick: why does this need a getter and setter?

jrose-apple commented on test/Frontend/skip-non-inlinable-function-bodies.swift:
```diff
@@ -0,0 +1,248 @@
+// 2. Emit the SIL for a module and check it against the expected strings in function bodies.
+//    If we're doing the right skipping, then the strings that are CHECK-NOT'd won't have been typechecked or SILGen'd.
+
+// RUN: %target-swift-frontend -emit-sil -O -experimental-skip-non-inlinable-function-bodies %s 2>&1 | %FileCheck %s --check-prefixes CHECK,CHECK-SIL-ONLY
```
Looks like you didn't end up using CHECK-SIL-ONLY.

jrose-apple commented on test/Frontend/skip-non-inlinable-function-bodies.swift:
```diff
@@ -0,0 +1,248 @@
+// 3. Emit the module interface while skipping non-inlinable function bodies. Check it against the same set of strings.
+
+// RUN: %target-swift-frontend -typecheck %s -enable-library-evolution -emit-module-interface-path %t/Module.skipping.swiftinterface -experimental-skip-non-inlinable-function-bodies
+// RUN: %FileCheck %s --check-prefixes CHECK,CHECK-INTERFACE-ONLY < %t/Module.skipping.swiftinterface
```
…or CHECK-INTERFACE-ONLY
